spec/index.adoc

```diff
@@ -560,7 +560,15 @@ In order to call a canister, the user makes a POST request to `/api/v2/canister/
 * `method_name` (`text`): Name of the canister method to call
 * `arg` (`blob`): Argument to pass to the canister method
 
-The HTTP response to this request has an empty body and HTTP status 202, or a HTTP error (4xx or 5xx). Paranoid agents should not trust this response, and use <<http-read-state,`read_state`>> to determine the status of the call.
+The HTTP response to this request can have the following responses:
+
+* 202 HTTP status with empty body. Implying the request was accepted by the IC for further processing. Users should use <<http-read-state,`read_state`>> to determine the status of the call.
+* 200 HTTP status with non-empty body. Implying an execution pre-processing error occurred. The body of the response contains more information about the IC specific error encountered. The body is a CBOR map with the following fields:
+** `reject_code` (`nat`): The reject code (see <<reject-codes>>).
+** `reject_message` (`text`): a textual diagnostic message.
+** `error_code` (`text`): an optional implementation-specific textual error code (see <<error-codes>>).
+* 4xx HTTP status for client errors (e.g. malformed request or canister not found). Retrying the request will likely have the same outcome.
+* 5xx HTTP status when the server has encountered an error or is otherwise incapable of performing the request. The request might succeed if retried at a later time.
```

Comment:
Would this include e.g. `not enough cycles`? In this case, I'd like to see specific documented cases that can be 
matched against, similar to the above `error_code` because  some errors can be treated in some way by the developer.

On second thought, why not also allow this for any non-200 error code? There will always be use cases for differentiating between different error conditions, and only using HTTP status codes feels too limiting

Reply:
`not enough cycles` would be 2xx error

> On second thought, why not also allow this for any non-200 error code? There will always be use cases for differentiating between different error conditions, and only using HTTP status codes feels too limiting

Not sure what you exactly mean by this .

Reply:
Some of the 4xx or 5xx status codes have multiple root causes, some of which are recoverable for the client, but at the moment they can only be differentiated with the error message, which may change at any point in time. If we had an error code for these too, I'd be much more comfortable to implement automatic recovery processes.

Reply:
ideally figuring out if it is recoverable or not the client should use only the status code of the 4xx or 5xx error and not the body. This is part of the HTTP spec to.

Reply:
Then can we please guarantee that one cause of error = one status code? Otherwise it's back to matching on the body since there are no guarantees - having an `error_code` field would make this a lot easier.

One use case I have in mind: a non-controller asks for the canister status. I want to show the user an explanation of controllers / nice error message. On the replica, I can match on HTTP 403, but on ic-ref, it returns 400, where I have to match on the body for "wrong sender" because anything could return a 400.

Similarly, AFAIK 503 only happens if a query happens to hit a replica that is behind, therefore I should retry my request, but who knows? It could be a problem where retrying does not help